### DISCUSSION TITLE
Remote containers in VSCodium

### DISCUSSION THREAD

**Discussion by Weissnix4711** [VSCodium/vscodium#1471] [General]
Hi,

I would like to use devcontainers in VSCodium. Is this possible?

The issues I've been facing so far seem to be that when the Remote Containers extension creates a container, it tries to install the closed source vscode server inside the container. It does this by getting the URL `https://update.code.visualstudio.com/commit:e04433b713b33cd229ae8fbb747c31997b7a5074/server-linux-x64/stable` where the commit hash is taken from the `product.json` file. The domain however does not seem to be configurable. It seems to be hard-coded into the extension.

So, obviously this VSCodium commit hash from my release isn't going to be found on Microsoft's servers. I therefore tried changing the `commit` option in product.json to the [hash of the commit which this VSCodium build is based on](https://github.com/microsoft/vscode/commit/784b0177c56c607789f9638da7b6bf3230d47a8c). This seems to work at first, as that buid can be found on the vscode update server, and is installed in the container.

However this leads to further issues as the `serverApplicationName` and `serverDataFolderName` dir names are also taken from the product.json file. Those directories obviously don't exist in the proprietary server release which has been installed in the container. So, I changed those values as well. `codium-server` becomes `code-server` and `.vscodium-server` becomes`.vscode-server`. This probably breaks god knows what else, but never mind.

Now the container finishes building installing and running whatever tests. That's great, but the workspace is still not opened. I get an error message which says `Failed to connect to the remote extension host server (Error: Connection error: Unauthorized client refused)`. I presume this is because of a version mismatch? I did make sure to use the commit hash which my VSCodium build is based on, but apparently this is not good enough?

Does anyone know if this can even be done, or am I wasting my time here? Is there a way to either install a different non-ms build of vscode server inside the container, or can anything be done to force code to connect to the incorrect server version regardless?

**Comment by jeanp413** [comment]
>  That's great, but the workspace is still not opened. I get an error message which says Failed to connect to the remote extension host server (Error: Connection error: Unauthorized client refused). I presume this is because of a version mismatch? I did make sure to use the commit hash which my VSCodium build is based on, but apparently this is not good enough?

It's not possble to connect vscodium to vscode-server, vscode from MS uses a Node native addon `vsda` (not bundled in code-oss) which is some kind of DRM so only MS vscode desktop/server can connect which each other [[1](https://parsiya.net/blog/2021-12-20-rce-in-visual-studio-codes-remote-wsl-for-fun-and-negative-profit/#your-editor-has-drm)]

**Comment by daiyam** [comment]
@Weissnix4711 You should use the assets `reh` which are `codium-server`.
And https://github.com/xaberus/vscode-remote-oss might be useful.
I will need to make times to iron out all of this...

[a comment by an automated account is not shown]

**Comment by dbrgn** [comment]
I'd be interested too in a way to use the devcontainers extension with the open source version of Code... @github-actions please un-stale.

**Comment by kikislater** [comment]
same for me

**Comment by GitMensch** [comment]
If I understand that correctly you'd have to do one of the following:
* adjust the devcontainer extension - to make the URL configurable; which won't work if that is a ms-licensed non-source-available extension
* create an own devcontainer extension that does the same steps than the one you refer to - which now is configurable and possibly default to vscodium
* use an existing remote extension and the devcontainer parts, possibly https://github.com/xaberus/vscode-remote-oss or https://github.com/jeanp413/open-remote-ssh

In any case there isn't anything vscodium can directly do about that (all necessary code is in, all binaries including reh are built and distributed), so I'm converting this to a discussion.

**Comment by s-light** [comment]
does anybody has any info / uses something that is working?

my college told me about devcontainers and it would make live easier for some of our devs in the team..

as iam the only one using vscodium (the rest uses VSC) its fine if i have to do some extra work to get it working ;-)

the goal is to *setup a devcontiainer config* that works for all of us ;-)

if one of you have any other good info on what is the overall thing that needs to be done please let me know..

i found https://github.com/xaberus/vscode-remote-oss/issues/2
but iam not quite sure that i really understand what is done in the bigger picture to get all this working...

**Reply by s-light** [reply]
ok it seems i have to dig a lot deeper into all of this..

found these to tutorials and will try theme..
- https://howtos.davidsebek.com/vscodium-containers.html
- https://www.myview.is/posts/tech/devcontainers/

**Reply by kiprasmel** [reply]
i have my own setup for developing thru docker containers, utilizing a different vscode ssh extension:
- https://github.com/jeanp413/open-remote-ssh

i've written about it here:
- https://blog.kipras.org/devel-core-git-macos#lightweight-linux-environment

in short, you have a docker container that runs sshd, and use the vscode extension to connect to it.

**Reply by daiyam** [reply]
@kiprasmel Thx. Great doc!

I think it would be good if your image in the docker hub (https://hub.docker.com/r/kipras/git-dev) is linked to its source (https://github.com/kiprasmel/infra/tree/master/local/devel-core-git) and explain its relation with https://github.com/kiprasmel/voidrice 😉

**Reply by craigphicks** [reply]
https://github.com/craigphicks/vscodium-docker-ide-demo combines
- VSCodium
- Open Remote - SSH [by jeanp413]
- Some extra code to put in a Dockerfile
- A bash script
- A README with a step by step guide.

A limitation is that it has only been tested with the containers on the same host machine running VSCodium. (Remote containers is a TODO).

**Reply by travankor** [reply]
I don't think any of these support arbitrary `devcontainer.json` for pulling in a reproducible environment or using published devcontainer templates. Or support distrobox integration described here https://distrobox.it/posts/integrate_vscode_distrobox/#from-flatpak

**Reply by craigphicks** [reply]
@travankor - re: supporting arbitrary devcontainer.json

The [Devcontainer CLI repo](https://github.com/devcontainers/cli/) has a [folder with examples](https://github.com/devcontainers/cli/tree/main/example-usage) of using `devcontains` cli interface to set up remote devcontainers which can be accessed by a client.

Currently it offers three examples:

- tool-vscode-server 
    - VS Code Server (official).  (The client would be a vscode instance.)
- tool-openvscode-server
    - [openvscode-server](https://github.com/gitpod-io/openvscode-server) (The client would be a browser)
- tool-vim-via-ssh
    - Vim via an SSH connection. SSH is used primarily to demonstrate how this could be achieved from other SSH supporting client tools. (The client would be a terminal).

(Note: The clients are not part of the example software).

All of them share the same action of creating a devcontainer from a `devcontainer.json` file, etc.  
Where the examples differ is installing different server software on the container, which each have their own protocols to communicate with their client.


A recent [pull](https://github.com/devcontainers/cli/pull/979) adds another example 

- tool-vscodium-server
    - VS Codium Remote Server.  (The client would be a vscodium instance.)

This is very similar to the *tool-vscode-server example, but with vscode-server replaced by vscodium-server, and where the the vscode client would be replaced by a vscodium client.  

The vscodium client is fairly easy to set up.
- ensure the a connection is possible between server and client
- install the remote oss (by xaberus)
- set the hostname and port, etc., in the remote oss settings
- connect

The [Devcontainer CLI repo](https://github.com/devcontainers/cli/) repo itself opens in a devcontainer.  It is set up to be able to build and run another inner devcontainer inside that outer container.  By running the outer container in VSCode, the outer container will automatically forward the ports to the localhost on your machine where the client is running - thus taking care of the "ensure the a connection is possible between server and client" task.

In a realistic use case omitting VSCode would require forwarding the ports in another way.

**Reply by remus-selea** [reply]
There's also [open-remote-distrobox](https://github.com/nerditation/open-remote-distrobox), which seems to be based on [vscode-remote-oss](https://github.com/xaberus/vscode-remote-oss) or at least inspired by it. I tried it out and was able to connect to a distrobox.

**Comment by aergic-dev** [comment]
Also see [this discussion](https://github.com/VSCodium/vscodium/discussions/2894)...released a new plugin for vscodium that supports devcontainers.

Edit: Newly released 0.3.0 now also enables devcontainer creation/use from a remote-ssh installation.

**Comment by s-h-a-d-o-w** [comment]
I just published a dev containers extension that works like the ones in VS Code or Cursor from a basic UX point of view. Have a `devcontainer.json` and you'll get a prompt to open in a container. Or go through the "remote" menu on the bottom left. No other extensions, tools, config or SSH needed.
https://open-vsx.org/extension/s-h-a-d-o-w/dev-containers-oss
